fuzz/ml-kem.c

```diff
@@ -0,0 +1,685 @@
+ */
+static uint8_t *consume_uint16t(const uint8_t *buf, size_t *len, uint16_t *val)
+{
+    uint8_t *buf_idx = (uint8_t *)buf;
+
+    if (*len < (sizeof(uint16_t) * 2))
+        return NULL;
+    /*
+     * Align the buffer to the next 16 bit boundary
+     */
+    buf_idx = (buf_idx + (sizeof(uint16_t) - 1));
+    buf_idx = (uint8_t *)((uintptr_t)buf_idx & (uintptr_t)(~(sizeof(uint16_t) - 1)));
+    *val = (uint16_t)*buf_idx;
+    /* advance the buffer pointer */
+    buf_idx += sizeof(uint16_t);
+    /* compute our remaining length */
+    *len -= (buf_idx - buf);
+    return buf_idx;
+}
+
+/**
+ * @brief Consumes an 8-bit unsigned integer from a buffer.
+ *
+ * This function extracts an 8-bit unsigned integer from the provided buffer,
+ * updates the buffer pointer, and adjusts the remaining length.
+ *
+ * @param buf  Pointer to the input buffer.
+ * @param len  Pointer to the size of the remaining buffer; updated after consumption.
+ * @param val  Pointer to store the extracted 8-bit value.
+ *
+ * @return Pointer to the updated buffer position after reading the value,
+ *         or NULL if the buffer does not contain enough data.
+ */
+static uint8_t *consume_uint8t(const uint8_t *buf, size_t *len, uint8_t *val)
+{
+    if (*len < sizeof(uint8_t))
+        return NULL;
+    *val = *buf;
+    *len -= sizeof(uint8_t);
+    return (uint8_t *)buf + 1;
+}
+
+/**
+ * @brief Selects a key type and size from a buffer.
+ *
+ * This function reads a key size value from the buffer, determines the
+ * corresponding key type and length, and updates the buffer pointer
+ * accordingly. If `only_valid` is set, it restricts selection to valid
+ * key sizes; otherwise, it includes some invalid sizes for testing.
+ *
+ * @param buf       Pointer to the buffer pointer; updated after reading.
+ * @param len       Pointer to the remaining buffer size; updated accordingly.
+ * @param keytype   Pointer to store the selected key type string.
+ * @param keylen    Pointer to store the selected key length.
+ * @param only_valid Flag to restrict selection to valid key sizes.
+ *
+ * @return 1 if a key type is successfully selected, 0 on failure.
+ */
+static int select_keytype_and_size(uint8_t **buf, size_t *len,
+                                   char **keytype, size_t *keylen,
+                                   int only_valid)
+{
+    uint16_t keysize;
+    uint16_t modulus = 6;
+
+    *buf = consume_uint16t(*buf, len, &keysize);
+    if (*buf == NULL)
+        return 0;
+
+    /*
+     * select from sizes
+     * ML-KEM-512, ML-KEM-768, and ML-KEM-1024
+     * also select some invalid sizes to trigger
+     * error paths
+     */
+    if (only_valid)
+        modulus = 3;
+
+    /*
+     * Note, keylens for valid values (cases 0-2)
+     * are taken based on input values from our unit tests
+     */
+    switch (keysize % modulus) {
+    case 0:
+        *keytype = "ML-KEM-512";
+        *keylen = 800;
```

Comment:
This is the public key size of ML-KEM-512. For ML-KEM, the private key size is always 2*public + 32, in this case 1632 if that's relevant.

Reply:
We do, and the updated PR uses those macros now.  We also randomly select invalid algorithms, as well as valid algorithms plus invalid lengths (see cases 3 and 4 in the switch statement in select_keytype_and_size()